api/rql/internal/predicate/string.go

```diff
@@ -0,0 +1,234 @@
+
+func (p *stringGlob) Unmarshal(input interface{}) error {
+	if !matcher.Array(matcher.Value("glob"))(input) {
+		return errz.MatchErrorf("must be formatted as ['glob', <glob_str>]")
+	}
+	array := input.([]interface{})
+	if len(array) > 2 {
+		return fmt.Errorf("must be formatted as ['glob', <glob_str>]")
+	}
+	if len(array) < 2 {
+		return fmt.Errorf("missing the glob")
+	}
+	globStr, ok := array[1].(string)
+	if !ok {
+		return fmt.Errorf("glob must be a string")
+	}
+	g, err := glob.Compile(globStr)
+	if err != nil {
+		return fmt.Errorf("invalid glob %v: %w", globStr, err)
+	}
+	p.gStr = globStr
+	p.g = g
+	return nil
+}
+
+func (p *stringGlob) EvalString(str string) bool {
+	return p.g.Match(str)
+}
+
+var _ = rql.StringPredicate(&stringGlob{})
+
+func StringRegex(r *regexp.Regexp) rql.StringPredicate {
+	return &stringRegex{
+		r: r,
+	}
+}
+
+type stringRegex struct {
+	r *regexp.Regexp
+}
+
+func (p *stringRegex) Marshal() interface{} {
+	return []interface{}{"regex", p.r.String()}
+}
+
+func (p *stringRegex) Unmarshal(input interface{}) error {
+	if !matcher.Array(matcher.Value("regex"))(input) {
+		return errz.MatchErrorf("must be formatted as ['regex', <regex_str>]")
+	}
+	array := input.([]interface{})
+	if len(array) > 2 {
+		return fmt.Errorf("must be formatted as ['regex', <regex_str>]")
+	}
+	if len(array) < 2 {
+		return fmt.Errorf("missing the regex")
+	}
+	regexStr, ok := array[1].(string)
+	if !ok {
+		return fmt.Errorf("regex must be a string")
+	}
+	r, err := regexp.Compile(regexStr)
+	if err != nil {
+		return fmt.Errorf("invalid regex %v: %w", regexStr, err)
+	}
+	p.r = r
+	return nil
+}
+
+func (p *stringRegex) EvalString(str string) bool {
+	return p.r.MatchString(str)
+}
+
+var _ = rql.StringPredicate(&stringRegex{})
+
+func StringEqual(s string) rql.StringPredicate {
+	return &stringEqual{
+		s: s,
+	}
+}
+
+type stringEqual struct {
+	s string
+}
+
+func (p *stringEqual) Marshal() interface{} {
+	return []interface{}{EQL, p.s}
+}
+
+func (p *stringEqual) Unmarshal(input interface{}) error {
+	if !matcher.Array(matcher.Value(EQL))(input) {
+		return errz.MatchErrorf("must be formatted as ['%v', <str>]", EQL)
+	}
+	array := input.([]interface{})
+	if len(array) > 2 {
+		return fmt.Errorf("must be formatted as ['%v', <str>]", EQL)
+	}
+	if len(array) < 2 {
+		return fmt.Errorf("missing the string")
+	}
+	s, ok := array[1].(string)
+	if !ok {
+		return fmt.Errorf("must provide a string")
+	}
+	p.s = s
+	return nil
+}
+
+func (p *stringEqual) EvalString(str string) bool {
+	return str == p.s
+}
+
+var _ = rql.StringPredicate(&stringEqual{})
+
+/*
+This is the main string predicate type
+*/
+
+type stringP struct {
+	internal.NonterminalNode
+}
+
+func String() rql.StringPredicate {
+	p := &stringP{
+		NonterminalNode: internal.NewNonterminalNode(
+			StringGlob(""),
+			StringRegex(nil),
+			StringEqual(""),
+		),
+	}
+	p.SetMatchErrMsg("must be formatted as either ['glob', <glob>], ['regex', <regex>], or ['=', <str>]")
```

Comment:
Why not use `EQL` here as well? You went to the effort of making it a constant in lines 123 and 127.

Reply:
Went back to just using "=" because comparisonOp caused some annoying test failures. The constant was originally so `Time`, `Numeric` would get the right comparisonOp; that security is not needed since string equal has a helper, `StringEqual` that provides that.

Reply:
Updated